File: sdk/kotlin/src/main/kotlin/com/seocheon/sdk/modules/EpochModule.kt
package com.seocheon.sdk.modules

import com.seocheon.sdk.constants.ChainConstants
import com.seocheon.sdk.infrastructure.ChainClient
import com.seocheon.sdk.types.EpochInfoResponse
import com.seocheon.sdk.types.QualificationResponse
import com.seocheon.sdk.types.WindowActivity
import com.seocheon.sdk.utils.EpochUtils
import kotlinx.serialization.json.*

/**
 * Epoch module for querying epoch/window state and activity qualification.
 */
class EpochModule(
    private val client: ChainClient,
) {

    /**
     * Returns current epoch and window information.
     */
    suspend fun getInfo(): EpochInfoResponse {
        val blockResp = client.getLatestBlock()
        val block = blockResp.jsonObject["block"]?.jsonObject
        val header = block?.get("header")?.jsonObject
        val blockHeight = header?.get("height")?.jsonPrimitive?.longOrNull ?: 0L

        val epochNumber = EpochUtils.computeEpoch(blockHeight)
        val windowNumber = EpochUtils.computeWindow(blockHeight)

        val epochStart = EpochUtils.epochStartBlock(epochNumber)
        val epochEnd = EpochUtils.epochEndBlock(epochNumber)
        val windowStart = EpochUtils.windowStartBlock(epochStart, windowNumber)
        val windowEnd = EpochUtils.windowEndBlock(windowStart)

        val epochProgress = if (epochEnd > epochStart)
            "%.2f".format((blockHeight - epochStart).toDouble() / (epochEnd - epochStart) * 100)
        else "0.00"

        val windowProgress = if (windowEnd >= windowStart)
            "%.2f".format((blockHeight - windowStart).toDouble() / ChainConstants.WINDOW_LENGTH * 100)
        else "0.00"

        return EpochInfoResponse(
            blockHeight = blockHeight,
            epochNumber = epochNumber,
            epochStartBlock = epochStart,
            epochEndBlock = epochEnd,
            epochProgress = epochProgress,
            windowNumber = windowNumber,
            windowStartBlock = windowStart,
            windowEndBlock = windowEnd,
            windowProgress = windowProgress,
            blocksUntilNextWindow = windowEnd - blockHeight,
            blocksUntilNextEpoch = epochEnd - blockHeight,
        )
    }

    /**
     * Queries reward qualification status for a node.
     */
    suspend fun getQualification(nodeId: String, epochNumber: Long? = null): QualificationResponse {
        val currentEpoch = epochNumber ?: EpochUtils.computeEpoch(getCurrentBlockHeight())

        val resp = client.queryRest("/seocheon/activity/v1/qualification/$nodeId?epoch_number=$currentEpoch")
        val obj = resp.jsonObject

        val windowDetail = obj["window_detail"]?.jsonArray?.map { item ->
            val w = item.jsonObject
            WindowActivity(
                windowNumber = w["window_number"]?.jsonPrimitive?.longOrNull ?: 0L,
                submissionCount = w["submission_count"]?.jsonPrimitive?.longOrNull ?: 0L,
                hasActivity = w["has_activity"]?.jsonPrimitive?.booleanOrNull ?: false,
            )
        } ?: emptyList()

        val activeWindows = obj["active_windows"]?.jsonPrimitive?.longOrNull
            ?: windowDetail.count { it.hasActivity }.toLong()
        val requiredWindows = ChainConstants.MIN_ACTIVE_WINDOWS
        val isQualified = activeWindows >= requiredWindows

        val elapsedWindows = obj["elapsed_windows"]?.jsonPrimitive?.longOrNull
            ?: ChainConstants.WINDOWS_PER_EPOCH
        val remainingNeeded = maxOf(0, requiredWindows - activeWindows)
        val remainingWindows = ChainConstants.WINDOWS_PER_EPOCH - elapsedWindows
        val canStillQualify = activeWindows + remainingWindows >= requiredWindows

        return QualificationResponse(
            epochNumber = currentEpoch,
            totalWindows = ChainConstants.WINDOWS_PER_EPOCH,
            elapsedWindows = elapsedWindows,
            activeWindows = activeWindows,
            requiredWindows = requiredWindows,
            isQualified = isQualified,
            remainingNeeded = remainingNeeded,
            canStillQualify = canStillQualify,
            windowDetail = windowDetail,
        )
    }

    private suspend fun getCurrentBlockHeight(): Long {
        val blockResp = client.getLatestBlock()
        val block = blockResp.jsonObject["block"]?.jsonObject
        val header = block?.get("header")?.jsonObject
        return header?.get("height")?.jsonPrimitive?.longOrNull ?: 0L
    }
}

File: sdk/csharp/src/Modules/EpochModule.cs
using Seocheon.Sdk.Constants;
using Seocheon.Sdk.Infrastructure;
using Seocheon.Sdk.Infrastructure.Signing;
using Seocheon.Sdk.Errors;
using Seocheon.Sdk.Types;
using Seocheon.Sdk.Utils;

namespace Seocheon.Sdk.Modules;

/// <summary>
/// Epoch module: query epoch/window state and qualification status.
/// </summary>
public sealed class EpochModule
{
    private readonly IChainClient _client;
    private readonly ISigningService _signer;

    internal EpochModule(IChainClient client, ISigningService signer)
    {
        _client = client;
        _signer = signer;
    }

    /// <summary>
    /// Returns current epoch and window information.
    /// </summary>
    public async Task<EpochInfoResponse> GetInfo(CancellationToken ct = default)
    {
        var block = await _client.GetLatestBlock(ct);
        var height = long.Parse(
            block.GetProperty("block").GetProperty("header").GetProperty("height").GetString() ?? "1"
        );

        var epoch = EpochUtils.ComputeEpoch(height);
        var window = EpochUtils.ComputeWindow(height);
        var epochStart = EpochUtils.EpochStartBlock(epoch);
        var epochEnd = EpochUtils.EpochEndBlock(epoch);
        var windowStart = EpochUtils.WindowStartBlock(epochStart, window);
        var windowEnd = EpochUtils.WindowEndBlock(windowStart);

        var epochElapsed = height - epochStart + 1;
        var epochTotal = ChainConstants.EpochBlocks;
        var windowElapsed = height - windowStart + 1;
        var windowTotal = ChainConstants.WindowBlocks;

        return new EpochInfoResponse(
            BlockHeight: height,
            EpochNumber: epoch,
            EpochStartBlock: epochStart,
            EpochEndBlock: epochEnd,
            EpochProgress: $"{epochElapsed}/{epochTotal}",
            WindowNumber: window,
            WindowStartBlock: windowStart,
            WindowEndBlock: windowEnd,
            WindowProgress: $"{windowElapsed}/{windowTotal}",
            BlocksUntilNextWindow: EpochUtils.BlocksUntilNextWindow(height),
            BlocksUntilNextEpoch: EpochUtils.BlocksUntilNextEpoch(height)
        );
    }

    /// <summary>
    /// Returns qualification status for a node in an epoch.
    /// </summary>
    /// <param name="nodeId">Node ID. Empty resolves from signer's agent address.</param>
    /// <param name="epochNumber">Epoch number. 0 means current epoch.</param>
    /// <param name="ct">Cancellation token.</param>
    public async Task<QualificationResponse> GetQualification(string nodeId = "", long epochNumber = 0, CancellationToken ct = default)
    {
        if (string.IsNullOrEmpty(nodeId))
            nodeId = await ResolveOwnNodeId(ct);

        if (epochNumber == 0)
            epochNumber = await ComputeCurrentEpoch(ct);

        var result = await _client.QueryRest(
            $"/seocheon/activity/v1/qualification/{nodeId}?epoch={epochNumber}",
            ct
        );

        var totalWindows = ChainConstants.WindowsPerEpoch;
        var requiredWindows = ChainConstants.MinActiveWindows;

        long elapsedWindows = 0;
        if (result.TryGetProperty("elapsed_windows", out var ew))
            long.TryParse(ew.GetString(), out elapsedWindows);

        ulong activeWindows = 0;
        if (result.TryGetProperty("active_windows", out var aw))
            ulong.TryParse(aw.GetString(), out activeWindows);

        var isQualified = activeWindows >= (ulong)requiredWindows;
        var remainingNeeded = Math.Max(0, requiredWindows - (long)activeWindows);
        var windowsLeft = totalWindows - elapsedWindows;
        var canStillQualify = isQualified || remainingNeeded <= windowsLeft;

        var windowDetail = new List<WindowActivity>();
        if (result.TryGetProperty("window_detail", out var wd))
        {
            foreach (var w in wd.EnumerateArray())
            {
                var wn = w.TryGetProperty("window_number", out var wnum)
                    ? long.Parse(wnum.GetString() ?? "0") : 0;
                ulong sc = 0;
                if (w.TryGetProperty("submission_count", out var scv))
                    ulong.TryParse(scv.GetString(), out sc);
                windowDetail.Add(new WindowActivity(wn, sc, sc > 0));
            }
        }

        return new QualificationResponse(
            EpochNumber: epochNumber,
            TotalWindows: totalWindows,
            ElapsedWindows: elapsedWindows,
            ActiveWindows: activeWindows,
            RequiredWindows: requiredWindows,
            IsQualified: isQualified,
            RemainingNeeded: remainingNeeded,
            CanStillQualify: canStillQualify,
            WindowDetail: windowDetail
        );
    }

    private async Task<string> ResolveOwnNodeId(CancellationToken ct)
    {
        var address = _signer.GetAddress();
        var result = await _client.QueryRest($"/seocheon/node/v1/node_by_agent/{address}", ct);
        return result.GetProperty("node").GetProperty("node_id").GetString()
               ?? throw SdkErrors.QueryFailed("Could not resolve node ID");
    }

    private async Task<long> ComputeCurrentEpoch(CancellationToken ct)
    {
        var block = await _client.GetLatestBlock(ct);
        var height = long.Parse(
            block.GetProperty("block").GetProperty("header").GetProperty("height").GetString() ?? "1"
        );
        return EpochUtils.ComputeEpoch(height);
    }
}
